diff --git a/src/terminal.c b/src/terminal.c
index 7909fdf..d76a3c4 100644
--- a/src/terminal.c
+++ b/src/terminal.c
@@ -463,6 +463,7 @@ main (int argc, char **argv)
 	GError *error = NULL;
 	char *working_directory;
 	int ret = EXIT_SUCCESS;
+	GtkSettings *settings;
 
 	setlocale (LC_ALL, "");
 
@@ -516,6 +517,12 @@ main (int argc, char **argv)
 
 	g_set_application_name (_("Terminal"));
 
+	settings = gtk_settings_get_default ();
+
+	g_object_set (G_OBJECT (settings),
+		"gtk-application-prefer-dark-theme", TRUE,
+		NULL);
+
 	/* Unset the these env variables, so they doesn't end up
 	 * in the factory's env and thus in the terminals' envs.
 	 */
