diff --git a/src/xfce-bg.c b/src/xfce-bg.c
index c969f57..ab58dca 100644
--- a/src/xfce-bg.c
+++ b/src/xfce-bg.c
@@ -347,6 +347,13 @@ xfce_bg_check_property_prefix (XfconfChannel *channel,
                                gchar *prefix) {
     gchar *property;
 
+    property = g_strconcat (prefix, "/last-screensaver-image", NULL);
+    if (xfconf_channel_has_property (channel, property)) {
+        g_free (property);
+        return TRUE;
+    }
+    g_free (property);
+
     property = g_strconcat (prefix, "/last-image", NULL);
     if (xfconf_channel_has_property (channel, property)) {
         g_free (property);
@@ -456,7 +463,11 @@ xfce_bg_load_from_xfconf (XfceBG *bg,
 
     /* Filename */
     g_free (property);
-    property = g_strconcat (prop_prefix, "/last-image", NULL);
+    property = g_strconcat (prop_prefix, "/last-screensaver-image", NULL);
+    if (!xfconf_channel_has_property (channel, property)) {
+        g_free (property);
+        property = g_strconcat (prop_prefix, "/last-image", NULL);
+    }
     if (!xfconf_channel_has_property (channel, property)) {
         g_free (property);
         property = g_strconcat (prop_prefix, "/image-path", NULL);
